### ISSUE TITLE
memory leak

### ISSUE THREAD

**Issue by jcs** [igrigorik/http-2#73]
Regarding https://github.com/ostinelli/net-http2/issues/7, I was able to reproduce the memory leak outside of `net-http2` with a separate client, so the leak appears to be with `http-2`.

My class for sending messages is below, though even with the hack to delete `@listeners` it still leaks. (It's been a while since I looked at this code, so I'm guessing that made it leak less.) In production sending a handful of messages every second over a persistent connection (Apple's APNS servers), the process balloons to 512mb in about 7 hours (which triggers it to be restarted).

This is with Ruby 2.24p230 on OpenBSD.  Any help troubleshooting this would be appreciated. 

````
require "http/2"
require "openssl"
require "resolv"
require "ostruct"

# monkeypatch to free up memory
module HTTP2
  module Emitter
    def delete_listeners
      if @listeners
        @listeners.each do |k,v|
          @listeners.delete(k)
        end
      end

      @listeners = nil
    end
  end
end

class HTTP2Client
  attr_accessor :hostname, :ssl_context, :ssl_socket, :tcp_socket, :h2_client,
    :h2_stream, :headers, :body, :done

  DRAFT = "h2"

  def initialize(hostname, ip = nil, ctx = nil)
    if !ctx
      ctx = OpenSSL::SSL::SSLContext.new
    end

    ctx.npn_protocols = [ DRAFT ]
    ctx.npn_select_cb = lambda do |protocols|
      DRAFT if protocols.include?(DRAFT)
    end

    self.ssl_context = ctx

    if !ip
      ip = Resolv.getaddress(hostname)
    end

    begin
      Timeout.timeout(10) do
        self.tcp_socket = TCPSocket.new(ip, 443)

        sock = OpenSSL::SSL::SSLSocket.new(self.tcp_socket, self.ssl_context)
        sock.sync_close = true
        sock.hostname = hostname
        sock.connect

        self.ssl_socket = sock
      end
    rescue Timeout::Error
      self.log(:error, nil, "timed out connecting to #{host[:ip]}")
      self.store_client(nil, dev)
      return nil
    end

    self.hostname = hostname

    self.h2_client = HTTP2::Client.new
    self.h2_client.on(:frame) do |bytes|
      self.ssl_socket.print bytes
      self.ssl_socket.flush
      nil
    end
  end

  def call(method, path, options = {})
    if self.h2_stream
      self.cleanup_stream
    end

    headers = (options[:headers] || {})
    headers.merge!({
      ":scheme" => "https",
      ":method" => method.to_s.upcase,
      ":path" => path,
    })

    headers.merge!("host" => self.hostname)

    if options[:body]
      headers.merge!("content-length" => options[:body].bytesize.to_s)
    else
      headers.delete("content-length")
    end

    self.h2_stream = self.h2_client.new_stream

    self.headers = {}
    self.h2_stream.on(:headers) do |hs_array|
      hs = Hash[*hs_array.flatten]
      self.headers.merge!(hs)
      nil
    end

    self.body = ""
    self.h2_stream.on(:data) do |d|
      self.body << d
      nil
    end

    self.done = false
    self.h2_stream.on(:close) do |d|
      self.done = true
      nil
    end

    if options[:body]
      self.h2_stream.headers(headers, :end_stream => false)
      self.h2_stream.data(options[:body], :end_stream => true)
    else
      self.h2_stream.headers(headers, :end_stream => true)
    end

    while !self.ssl_socket.closed? && !self.ssl_socket.eof?
      data = self.ssl_socket.read_nonblock(1024)

      begin
        self.h2_client << data
      rescue => e
        self.ssl_socket.close
        raise e
      end

      if self.done
        break
      end
    end

    if self.done
      self.cleanup_stream
      return OpenStruct.new(:status => self.headers[":status"],
        :headers => self.headers, :body => self.body)
    else
      return nil
    end
  end

  def close
    begin
      if self.h2_client
        self.h2_client.close
      end
    rescue
    end

    begin
      if self.ssl_socket
        self.ssl_socket.close
      end
    rescue
    end

    begin
      if self.tcp_socket
        self.tcp_socket.close
      end
    rescue
    end
  end

  def cleanup_stream
    if self.h2_stream
      self.h2_stream.delete_listeners
    end

    begin
      self.h2_stream.close
    rescue
    end
  end
end
````

**Comment by ostinelli** [comment]
Thank you @jcs. FYI, an issue has been opened on Apnotic (https://github.com/ostinelli/apnotic/issues/37) regarding a potential memory leak, and it looks like it may be linked to your newly opened issue in the underlying Http2 gem (https://github.com/igrigorik/http-2/issues/73).

I think that I've found the culprit and have a patch for it.

We can confirm that there currently is a memory leak by running the following code against [NetHttp2 v0.14.0](https://github.com/ostinelli/net-http2/tree/0.14.0) with [Http2 v0.8.2](https://github.com/igrigorik/http-2/tree/v0.8.2):

```ruby
require 'net-http2'

COUNT = 1000

i = 0
c = NetHttp2::Client.new("https://http2.akamai.com/")

pid = Process.pid

while i < COUNT
  _res = c.call(:get, "/?#{Time.now.to_i}")

  rss = `ps -eo pid,rss | grep #{pid} | awk '{print $2}'`.to_i
  puts "rss: #{rss} - live objects: #{GC.stat[:heap_live_slots]}"

  GC.start
  sleep 0.5
  i += 1
end
```

The output is:

```
rss: 23404 - live objects: 95690
rss: 23416 - live objects: 50309
rss: 23416 - live objects: 50445
rss: 23416 - live objects: 50594
rss: 23416 - live objects: 50735
rss: 23416 - live objects: 50855
rss: 23416 - live objects: 51016
rss: 23420 - live objects: 51156
rss: 23420 - live objects: 51277
rss: 23420 - live objects: 51449
rss: 23428 - live objects: 51578
rss: 23432 - live objects: 51711

[omissis]

rss: 35772 - live objects: 137244
rss: 35784 - live objects: 137349
rss: 35812 - live objects: 137493
rss: 35828 - live objects: 137618
rss: 35848 - live objects: 137759
rss: 35884 - live objects: 137879
rss: 35944 - live objects: 138040
rss: 35972 - live objects: 138181
rss: 36012 - live objects: 138313
rss: 36024 - live objects: 138483
rss: 36048 - live objects: 138627
rss: 36072 - live objects: 138736

[omissis]

rss: 45492 - live objects: 188593
rss: 45520 - live objects: 188746
rss: 45572 - live objects: 188854
rss: 45628 - live objects: 189015
rss: 45692 - live objects: 189156
rss: 45708 - live objects: 189324
rss: 45720 - live objects: 189449
rss: 45772 - live objects: 189578
rss: 45796 - live objects: 189710
rss: 45808 - live objects: 189871
rss: 45844 - live objects: 190048
rss: 45868 - live objects: 190132
```

It therefore seems that there is a leak: I'm explicitly calling `GC.start` after every HTTP stream request, and yet both the rss and the heap live slots keep on increasing linearly. Let's look deeper and follow the methodology described [in this blog post](http://blog.skylight.io/hunting-for-leaks-in-ruby/). This is the modified test code:

```ruby
require 'net-http2'
require 'objspace'

ObjectSpace.trace_object_allocations_start

COUNT = 1000
DUMPS = 3

i = 0
j = 0

c = NetHttp2::Client.new("https://http2.akamai.com/")

while j < DUMPS
  while i < COUNT
    puts "#{j}:#{i}"

    _res = c.call(:get, "/?#{Time.now.to_i}")

    sleep 0.5
    i += 1
  end

  file_name = "heap-#{j}.dump"
  file      = File.open(file_name, 'w')

  GC.start
  ObjectSpace.dump_all(output: file)
  file.close

  i = 0
  j += 1
end
```

Running the above code generates three dumps: `heap-0.dump`, `heap-1.dump` and `heap-2.dump`. By diffing these files as described in the methodology referenced above, I get:

```
$ ruby diff.rb heap-0.dump heap-1.dump heap-2.dump
Leaked 29991 ARRAY objects of size 0/1199640 at: /Users/roberto/.rvm/gems/ruby-2.3.1@net-http2/gems/http-2-0.8.2/lib/http/2/emitter.rb:41
Leaked 19996 STRING objects of size 372967/1127840 at: /Users/roberto/.rvm/gems/ruby-2.3.1@net-http2/gems/http-2-0.8.2/lib/http/2/huffman.rb:33
Leaked 13994 STRING objects of size 219921/700713 at: /Users/roberto/workspace/net-http2/lib/net-http2/stream.rb:49
Leaked 8997 DATA objects of size 0/743752 at: /Users/roberto/.rvm/gems/ruby-2.3.1@net-http2/gems/http-2-0.8.2/lib/http/2/connection.rb:638
Leaked 8992 DATA objects of size 0/743272 at: /Users/roberto/.rvm/gems/ruby-2.3.1@net-http2/gems/http-2-0.8.2/lib/http/2/connection.rb:637
Leaked 6000 DATA objects of size 0/720000 at: /Users/roberto/.rvm/gems/ruby-2.3.1@net-http2/gems/http-2-0.8.2/lib/http/2/stream.rb:88
Leaked 5999 DATA objects of size 0/575920 at: /Users/roberto/.rvm/gems/ruby-2.3.1@net-http2/gems/http-2-0.8.2/lib/http/2/emitter.rb:41
Leaked 5000 STRING objects of size 7044/200000 at: /Users/roberto/.rvm/gems/ruby-2.3.1@net-http2/gems/http-2-0.8.2/lib/http/2/buffer.rb:22
Leaked 2999 DATA objects of size 0/239920 at: /Users/roberto/.rvm/gems/ruby-2.3.1@net-http2/gems/http-2-0.8.2/lib/http/2/stream.rb:89
Leaked 2999 IMEMO objects of size 0/119960 at: /Users/roberto/.rvm/gems/ruby-2.3.1@net-http2/gems/http-2-0.8.2/lib/http/2/connection.rb:637
Leaked 2998 IMEMO objects of size 0/119920 at: /Users/roberto/.rvm/gems/ruby-2.3.1@net-http2/gems/http-2-0.8.2/lib/http/2/connection.rb:638
Leaked 2998 OBJECT objects of size 0/407728 at: /Users/roberto/.rvm/gems/ruby-2.3.1@net-http2/gems/http-2-0.8.2/lib/http/2/connection.rb:629
Leaked 2998 HASH objects of size 0/2086512 at: /Users/roberto/.rvm/gems/ruby-2.3.1@net-http2/gems/http-2-0.8.2/lib/http/2/emitter.rb:41
Leaked 2998 ARRAY objects of size 0/119920 at: /Users/roberto/.rvm/gems/ruby-2.3.1@net-http2/gems/http-2-0.8.2/lib/http/2/stream.rb:86
Leaked 2000 DATA objects of size 0/184000 at: /Users/roberto/workspace/net-http2/lib/net-http2/stream.rb:60
Leaked 2000 DATA objects of size 0/184000 at: /Users/roberto/workspace/net-http2/lib/net-http2/stream.rb:48
Leaked 1999 DATA objects of size 0/207872 at: /Users/roberto/.rvm/gems/ruby-2.3.1@net-http2/gems/http-2-0.8.2/lib/http/2/emitter.rb:21
Leaked 1998 DATA objects of size 0/215784 at: /Users/roberto/.rvm/gems/ruby-2.3.1@net-http2/gems/http-2-0.8.2/lib/http/2/connection.rb:634
Leaked 1997 DATA objects of size 0/183712 at: /Users/roberto/workspace/net-http2/lib/net-http2/stream.rb:70
Leaked 1000 STRUCT objects of size 0/40000 at: /Users/roberto/workspace/net-http2/lib/net-http2/stream.rb:13
Leaked 1000 DATA objects of size 0/176000 at: /Users/roberto/workspace/net-http2/lib/net-http2/stream.rb:12
Leaked 1000 STRING objects of size 11692044/11817000 at: /Users/roberto/workspace/net-http2/lib/net-http2/stream.rb:8
Leaked 1000 HASH objects of size 0/1416000 at: /Users/roberto/workspace/net-http2/lib/net-http2/stream.rb:7
Leaked 1000 OBJECT objects of size 0/104000 at: /Users/roberto/workspace/net-http2/lib/net-http2/client.rb:68
Leaked 1000 ARRAY objects of size 0/40000 at: /Users/roberto/workspace/net-http2/lib/net-http2/stream.rb:13
Leaked 1000 STRING objects of size 20000/40000 at: /Users/roberto/workspace/net-http2/lib/net-http2/request.rb:27
Leaked 1000 HASH objects of size 0/40000 at: /Users/roberto/workspace/net-http2/lib/net-http2/request.rb:17
Leaked 1000 HASH objects of size 0/232000 at: /Users/roberto/.rvm/gems/ruby-2.3.1@net-http2/gems/http-2-0.8.2/lib/http/2/connection.rb:298
Leaked 999 OBJECT objects of size 0/95904 at: /Users/roberto/workspace/net-http2/lib/net-http2/client.rb:37
Leaked 999 HASH objects of size 0/231768 at: /Users/roberto/workspace/net-http2/lib/net-http2/request.rb:14
Leaked 999 STRING objects of size 11988/39960 at: test.rb:18
Leaked 998 STRING objects of size 2994/39920 at: /Users/roberto/workspace/net-http2/lib/net-http2/request.rb:23
Leaked 41 STRING objects of size 1043/2501 at: test.rb:27
Leaked 22 ARRAY objects of size 0/880 at: /Users/roberto/.rvm/gems/ruby-2.3.1@net-http2/gems/http-2-0.8.2/lib/http/2/compressor.rb:180
Leaked 1 ARRAY objects of size 0/40 at: /Users/roberto/.rvm/gems/ruby-2.3.1@net-http2/gems/http-2-0.8.2/lib/http/2/framer.rb:180
Leaked 1 DATA objects of size 0/224 at: /Users/roberto/.rvm/rubies/ruby-2.3.1/lib/ruby/2.3.0/openssl/buffering.rb:178
Leaked 1 OBJECT objects of size 0/40 at: /Users/roberto/.rvm/rubies/ruby-2.3.1/lib/ruby/2.3.0/openssl/buffering.rb:178
Leaked 1 STRING objects of size 16/40 at: /Users/roberto/.rvm/rubies/ruby-2.3.1/lib/ruby/2.3.0/openssl/buffering.rb:178
Leaked 1 HASH objects of size 0/232 at: /Users/roberto/.rvm/gems/ruby-2.3.1@net-http2/gems/http-2-0.8.2/lib/http/2/stream.rb:222


Total Size: 12328017/24396974
```

@igrigorik, as you can see most of those seem to come from the `Http2` gem. It may very well be that NetHttp2 is not releasing objects, but I didn't find anywhere where this might happen (or I might just have missed it!).

I've therefore looked at Http2's [connection.rb](https://github.com/igrigorik/http-2/blob/v0.8.2/lib/http/2/connection.rb) and I saw you are keeping a reference to the opened streams by adding them to the instance variable `@streams` [here](https://github.com/igrigorik/http-2/blob/v0.8.2/lib/http/2/connection.rb#L640). However, I don't seem to see anywhere these references being released once a stream is closed.

Therefore I've patched Http2's `connection.rb` to release the reference once a stream gets closed, editing the code from (see [here](https://github.com/igrigorik/http-2/blob/master/lib/http/2/connection.rb#L633)):

```ruby
stream.once(:close)  { @active_stream_count -= 1 }
```

to:

```ruby
stream.once(:close) do
  @streams.delete id
  @active_stream_count -= 1
end
```

After applying this patch, re-running the first code shows that rss and the heap live slots are now stable:

```
rss: 22852 - live objects: 95634
rss: 23008 - live objects: 50298
rss: 23012 - live objects: 50296
rss: 23012 - live objects: 50318
rss: 23100 - live objects: 50286
rss: 23100 - live objects: 50308
rss: 23100 - live objects: 50308
rss: 23100 - live objects: 50290
rss: 23100 - live objects: 50323
rss: 23160 - live objects: 50312
rss: 23160 - live objects: 50294
rss: 23160 - live objects: 50316

[omissis]

rss: 23468 - live objects: 50513
rss: 23468 - live objects: 50484
rss: 23468 - live objects: 50477
rss: 23468 - live objects: 50547
rss: 23468 - live objects: 50476
rss: 23468 - live objects: 50481
rss: 23468 - live objects: 50491
rss: 23468 - live objects: 50504
rss: 23468 - live objects: 50461
rss: 23468 - live objects: 50495
rss: 23468 - live objects: 50475
rss: 23468 - live objects: 50557

```

Furthermore, performing the heap dump analysis (after applying the patch) I get:

```
$ ruby diff.rb heap-0.dump heap-1.dump heap-2.dump
Leaked 40 STRING objects of size 1002/2390 at: test.rb:27
Leaked 1 HASH objects of size 0/232 at: test.rb:27


Total Size: 1002/2622
```

We're down to only 2 lines, both from the `test.rb` code itself (and not the underlying libraries). Hence, these two results combined seem to show that the memory leak is gone.

@igrigorik can you please confirm that this patch is desired, and that it doesn't break any of your tests?

**Comment by ostinelli** [comment]
This might be related to a possible memory leak that I traced down to `connection.rb` and described in https://github.com/ostinelli/net-http2/issues/7.

TL;DR: can you please patch `connection.rb` to release the reference once a stream gets closed, editing the code from (see [here](https://github.com/igrigorik/http-2/blob/master/lib/http/2/connection.rb#L633)):

```ruby
stream.once(:close)  { @active_stream_count -= 1 }
```

to:

```ruby
stream.once(:close) do
  @streams.delete id
  @active_stream_count -= 1
end
```

And see if this solves your issue?

Still, this needs vetting from @igrigorik.

**Comment by jcs** [comment]
@ostinelli I'm running that in production now.  I'll report back on memory usage later.

**Comment by jcs** [comment]
After about 15 hours, the memory usage hasn't moved a whole lot.  I think this has solved my problem.  Thanks @ostinelli!

````
USER       PID %CPU %MEM   VSZ   RSS TT  STAT  STARTED       TIME COMMAND
...       5419 19.0  1.8 183100 153456 p1  I      6:25PM    1:29.22 ruby22
````

````
USER       PID %CPU %MEM   VSZ   RSS TT  STAT  STARTED       TIME COMMAND
...       5419 21.3  2.3 219436 188812 p1  R     Wed06PM  283:24.71 ruby22
````

**Comment by ostinelli** [comment]
Great. @igrigorik has raised some good questions regarding implications of removing stream references, let's wait for him to vet and hopefully patch this. Being a memory leak I'm looking forward to push a fix in production :)

**Comment by jcs** [comment]
After a day and a half, it looks like memory usage is certainly way down from what it used to be, but it's still leaking somewhere at a slow rate.  I'm putting my `@listeners` deletion monkey patch back in and running with it now to see if both of these fixes make it stay low.

````
USER       PID %CPU %MEM   VSZ   RSS TT  STAT  STARTED       TIME COMMAND
...       5419 21.8  2.9 271396 242916 p1- I     Wed06PM  769:18.70 ruby22 /d/rails/pushover/script/send_messages
````

**Comment by ostinelli** [comment]
@jcs as per the memory leak research in https://github.com/ostinelli/net-http2/issues/7 I don't see any anymore, after this patch. I don't see why your listeners patch could work, since these are related to streams that are now freed (hence garbage collected).

That being said, please keep us updated.

**Comment by jcs** [comment]
Yeah, must have been a red herring when I first started looking at this stuff.

````
USER       PID %CPU %MEM   VSZ   RSS TT  STAT  STARTED       TIME COMMAND
....      8126 13.8  2.6 241688 213956 p0- R     Fri10AM  523:03.10 ruby22 
````

**Comment by ostinelli** [comment]
Looks good then.

@igrigorik what are your thoughts on this? My proposed change currently breaks a test:

```
Failures:

  1) HTTP2::Client push should auto RST_STREAM promises against locally-RST stream
     Failure/Error: expect(frame[:type]).to eq :rst_stream
     
       expected: :rst_stream
            got: :goaway
     
       (compared using ==)
     
       Diff:
       @@ -1,2 +1,2 @@
       -:rst_stream
       +:goaway
       
     # ./spec/client_spec.rb:85:in `block (4 levels) in <top (required)>'
     # ./lib/http/2/connection.rb:140:in `goaway'
     # ./lib/http/2/connection.rb:656:in `connection_error'
     # ./lib/http/2/connection.rb:281:in `receive'
     # ./spec/client_spec.rb:89:in `block (3 levels) in <top (required)>'

Finished in 0.79455 seconds (files took 0.27817 seconds to load)
339 examples, 1 failure

Failed examples:

rspec ./spec/client_spec.rb:78 # HTTP2::Client push should auto RST_STREAM promises against locally-RST stream
```

**Comment by igrigorik** [comment]
/me summons @kazuho for some sage implementation advice... :-)

Context from https://github.com/ostinelli/net-http2/issues/7: 

> I've therefore looked at Http2's connection.rb and I saw you are keeping a reference to the opened streams by adding them to the instance variable @streams here. However, I don't seem to see anywhere these references being released once a stream is closed.
> 
> Therefore I've patched Http2's connection.rb to release the reference once a stream gets closed, editing the code from (see here):
> 
> stream.once(:close)  { @active_stream_count -= 1 }
> to:
> 
> stream.once(:close) do
>   @streams.delete id
>   @active_stream_count -= 1
> end

---

> You're right, that's an oversight on our part.. we should be harvesting old streams. The actual spot/time on when to do so might require some experimentation though as h2 does allow us to receive frames on a closed stream in some cases. That said, we can start with something simple (as suggested above) and iterate from there.

@kazuho any recommendations for how to implement this, based on your experience with h2o?

**Comment by ostinelli** [comment]
Any news on this? This was unfortunately generating out of memory errors and had to manage, so we're currently using a patched version of `http-2`. Maybe a release with the simple suggestion I'm making and then a better one once @kazuho can provide some feedback? :)

**Comment by igrigorik** [comment]
Paging @kazuho @icing once more -- fingers crossed. Do you guys have any recommendations for best practices on releasing streams? :)

@ostinelli yeah, apologies about the delay. I'll push out an update this weekend if we don't hear back.

**Comment by ostinelli** [comment]
@igrigorik no worries - thank you for your great work.

**Comment by igrigorik** [comment]
Hey guys. Any chance you can try this patch in your implementations: https://github.com/igrigorik/http-2/pull/88

If that looks good, I'll merge and push an update to rubygems.

**Comment by icing** [comment]
hey @igrigorik, sorry for the late reply. Most of what Apache does in this regard is done by the nghttp2 lib. However, I had a discussion with @tatsuhiro-t once while checking results from h2spec. Certain expectations of it we did not meet, because it meant keeping information about closed streams around forever.

This is, if I understand it correctly, what this ticket is about. I think, the client should RST any promises against an unknown stream, as if that stream had been closed once. With a local close, those frames might be incoming for at least 2 RTTs, in practise much more due to buffering and it should not be a connection error, I think.

**Comment by igrigorik** [comment]
@icing thanks, that makes sense! Based on my research this weekend, seems like many implementations simply purge the closed stream and ignore the in-flight implications.. *shrug*. I'm hoping https://github.com/igrigorik/http-2/pull/88 might be a good compromise.

**Comment by kazuho** [comment]
@igrigorik Sorry for not responding. I knew that something was going on, but I assumed it was not an issue that was pinging me.

I'd assume that the changes in #88 would be more than enough, but the ideal way is to retain the information of _N_ closed streams, where N is calculated as _MAX_CONCURRENT_STREAMS - number-of-non-closed-streams_. Note that you'd need to calculate N for pushed and pulled streams separately.

**Comment by igrigorik** [comment]
@kazuho ah, interesting.. thanks for the tip! I guess that approach would enable a tighter bound on number of streams to keep a reference to. For first run at this, I'll keep #88 and I'll add a note to investigate your approach in a followup.

**Comment by ostinelli** [comment]
@kazuho this is interesting indeed. @igrigorik I'll try #88 asap, thank you!

**Comment by igrigorik** [comment]
@ostinelli any updates on the patch? If it works, would love to merge and push it live.

**Comment by ostinelli** [comment]
@igrigorik I'm currently using a patched version of http-2, and don't have the chance of pushing anything to prod before a week unfortunately. @jcs were you able to try @igrigorik's patch?

**Comment by jcs** [comment]
@ostinelli I am running it in production now.  No apparent regressions, though memory usage will take a while to compare.

**Comment by ostinelli** [comment]
Sounds great @jcs, thank you. @igrigorik I'd wait 1 week and see @jcs' memory usage on next Monday Dec 19th. If all is ok I'd push by then, which is around the time I will be able to push to production with this change. Would that work for you?

**Comment by igrigorik** [comment]
sgtm, thanks guys!

**Comment by jcs** [comment]
Memory usage is still reasonable after 48 hours.

````
USER       PID %CPU %MEM   VSZ   RSS TT  STAT  STARTED       TIME COMMAND
....     27033 21.2  3.0 267268 248136 p1- I     Mon10AM  897:08.06 ruby22
````

**Comment by ostinelli** [comment]
If @jcs confirms I think this can be released?

**Comment by igrigorik** [comment]
Merged and pushed 0.8.3 to rubygems. Closing.

Thanks everyone for helping track this one down!
